internal/controllers/machineset/machineset_controller_status.go

```diff
@@ -0,0 +1,350 @@
+	for _, machine := range machines {
+		if v1beta2conditions.IsTrue(machine, clusterv1.MachineReadyV1Beta2Condition) {
+			readyReplicas++
+		}
+		if v1beta2conditions.IsTrue(machine, clusterv1.MachineAvailableV1Beta2Condition) {
+			availableReplicas++
+		}
+		if v1beta2conditions.IsTrue(machine, clusterv1.MachineUpToDateV1Beta2Condition) {
+			upToDateReplicas++
+		}
+	}
+
+	if ms.Status.V1Beta2 == nil {
+		ms.Status.V1Beta2 = &clusterv1.MachineSetV1Beta2Status{}
+	}
+
+	ms.Status.V1Beta2.ReadyReplicas = ptr.To(readyReplicas)
+	ms.Status.V1Beta2.AvailableReplicas = ptr.To(availableReplicas)
+	ms.Status.V1Beta2.UpToDateReplicas = ptr.To(upToDateReplicas)
+}
+
+func setScalingUpCondition(_ context.Context, ms *clusterv1.MachineSet, machines []*clusterv1.Machine, bootstrapObjectNotFound, infrastructureObjectNotFound, getAndAdoptMachinesForMachineSetSucceeded bool) {
+	// If we got unexpected errors in listing the machines (this should never happen), surface them.
+	if !getAndAdoptMachinesForMachineSetSucceeded {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  clusterv1.MachineSetScalingUpInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+		return
+	}
+
+	// Surface if .spec.replicas is not yet set (this should never happen).
+	if ms.Spec.Replicas == nil {
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:   clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status: metav1.ConditionUnknown,
+			Reason: clusterv1.MachineSetScalingUpWaitingForReplicasSetV1Beta2Reason,
+		})
+		return
+	}
+
+	desiredReplicas := *ms.Spec.Replicas
+	if !ms.DeletionTimestamp.IsZero() {
+		desiredReplicas = 0
+	}
+	currentReplicas := ms.Status.Replicas
+
+	missingReferencesMessage := calculateMissingReferencesMessage(ms, bootstrapObjectNotFound, infrastructureObjectNotFound)
+
+	if currentReplicas >= desiredReplicas {
+		var message string
+		if missingReferencesMessage != "" {
+			message = fmt.Sprintf("Scaling up can't happen %s", missingReferencesMessage)
+		}
+		v1beta2conditions.Set(ms, metav1.Condition{
+			Type:    clusterv1.MachineSetScalingUpV1Beta2Condition,
+			Status:  metav1.ConditionFalse,
+			Reason:  clusterv1.MachineSetNotScalingUpV1Beta2Reason,
+			Message: message,
+		})
+		return
+	}
+
+	// Scaling up.
+	v1beta2conditions.Set(ms, metav1.Condition{
+		Type:   clusterv1.MachineSetScalingUpV1Beta2Condition,
+		Status: metav1.ConditionTrue,
+		Reason: clusterv1.MachineSetScalingUpV1Beta2Reason,
+		// Message: message,
+		Message: bulletMessage(
+			fmt.Sprintf("Scaling up from %d to %d replicas", currentReplicas, desiredReplicas),
+			aggregateMachinesFunc(
+				machines,
+				func(machine *clusterv1.Machine) bool {
+					return machine.Status.NodeRef == nil && time.Since(machine.GetCreationTimestamp().Time) >= time.Minute*30
+				},
+				"not reporting a .status.nodeRef by more than 30 minutes",
+			),
```

Comment:
xref: https://github.com/kubernetes-sigs/cluster-api/pull/11278#discussion_r1806478927

I think this should this be dropped according to the discussion? Should surface in MachinesReady anyway.